Topics/Image Processing/Convolution/Convolution.pde

```diff
@@ -1,31 +1,77 @@
 /**
  * Convolution
- * by Daniel Shiffman.  
- * 
- * Applies a convolution matrix to a portion of an image. Move mouse to 
- * apply filter to different parts of the image.
+ * by Daniel Shiffman.
+ *
+ * Applies a convolution matrix to a portion of an image. Move mouse to
+ * apply filter to different parts of the image. Click mouse to cycle
+ * through different effects (kernels).
  */
 
 PImage img;
+int effect = 0;
 int w = 120;
 
-// It's possible to convolve the image with many different 
-// matrices to produce different effects. This is a high-pass 
-// filter; it accentuates the edges. 
-float[][] matrix = { { -1, -1, -1 },
-                     { -1,  9, -1 },
-                     { -1, -1, -1 } }; 
+// It's possible to convolve the image with many different
+// matrices to produce different effects.
+String[] effect_names = {
+  "Identity (no change)",
+  "Darken",
+  "Lighten",
+  "Sharpen",
+  "Sharpen More",
+  "Box Blur",
+  "Edge Detect",
+  "Emboss"
+};
+
+float[][][] matrix = { { { 0, 0, 0 },    // Identity
```

Comment:
I'd suggest declaring the each kernel as its own separate array. It will be more verbose but will make things more legible and understandable for beginners.

Like so:

```Java
// no changes
float [][] identity = {{ 0, 0, 0 },
                       { 0, 1, 0 },
                       { 0, 0, 0 }};

float [][] darken = { { 0,   0, 0 },
                      { 0, 0.5, 0 },
                      { 0,   0, 0 } }

// etc.

float[][][] matrix = { identity, darken, ...
```